### DISCUSSION TITLE
Add `/subject` endpoint

### DISCUSSION THREAD

**Discussion by claymcleod** [CBIIT/ccdi-federation-api#7] [API Design Discussion]
## Overview

To facilitate the querying of subject data (both line level and summary), I propose the introduction of a `/subject` endpoint and subsequent namespace for the API.

## Approach

- **`GET /subject`.** This is the primary endpoint for retrieving line-level and summary data. Where allowed, this endpoint would return all subjects and available fields for all line-level data. Support will be provided for only returning particular fields of interest if desired. The data elements that are returned (and, subsequently, are available as filters) would be an ever growing list of metadata elements as defined by the harmonization subgroup. Note that, for sites that cannot provide line-level data, responses will be limited to the aggregation metrics such as total number of samples.
- **`GET /subject/<id>`.** Returns the subject object for a particular id (id here refers to the `id` key in the subject object, which is a unique integer primary key that is specific to the source server's database). For sites that only return line-level data, these endpoints will always return a 404.
- **`GET /subject/by/<field>/count`.** When aggregated counts are desired to be grouped by a particular field, users can use this endpoint to retrieve dynamically computed subject counts grouped by a particular data element. This should enable all available counting queries _besides_ just getting the total number of subjects contained in the server. Further, the API path is suffixed by `/count` to allow for other aggregating actions in the future.
- **`GET /metadata/fields/subject`.** Lists details about which metadata fields are supported by this server.

### Considerations

- Great consideration should be given to the fact that, in focus groups, the general consensus from the patient community appears to be that they prefer humanizing terms like "participant" or "patient" to the more general terms such as "subject" or "donor". That said, there is also a responsibility from the community to use a word that encapsulates a broad range of sources from which samples may be derived (including PDX and cell lines). To that end, the compromise I've struck in this proposal is to keep the noun name general enough to include all sources from which samples may be derived (`subject`) while allowing the designation of a "subject kind" to include designation of participants.
- Consideration was given to the fact whether the `/subject/<id>` endpoint should return a 403 ("Forbidden") or a 404 ("Not Found"). To that end, we discussed that (a) if it was _possible_ from the API perspective to return some results for this endpoint, but only when the user was authorized to access the data, then a 403 ("Forbidden") was the correct response, and (b) if it was _never possible_ to return results for this endpoint, then the correct response was a 404 ("Not Found"). Since there are no plans for authorization in v0 of the API, all responses fall into the latter category.

## Subject Retrieval

We'll start with the introduction of the `/subject/<id>` endpoint, which facilitates the retrieval of metadata regarding a single subject from the API. Below are definitions of [TypeScript](https://www.typescriptlang.org/) interfaces and types for each of the described entities in the responses. These provide more precise definitions for what keys and values are required in each response.

### Foundational Types

Before we dive into subject definitions, we'll need to define an `integer` type for general use throughout the specification.

```ts
/**
 * Defines an integer number in the range of -2^31 to 2^31 - 1.
 */
export type integer = number;
```

### `Subject`

With an `integer` now defined, we can begin with a formal specification of a proposed `Subject` model:

```ts
/**
 * A subject.
 */
export interface Subject {
  /**
   * A site-specific assigned id that will always correspond to this subject.
   * Note that these ids explicitly are not unique across the federation
   * ecosystem—uniqueness is only expected within a particular server or
   * implementation.
   */
  id: integer;

  /**
   * The primary name or identifier for a subject used within the source server.
   * This identifier should *ALWAYS* be included in the `identifiers` key under
   * metadata, should that object exist.
   */
  name: string;

  /**
   * The kind of subject.
   */
  kind: SubjectKind;

  /**
   * Metadata associated with the subject.
   */
  metadata?: SubjectMetadata;
}
```

Notably, a subject is comprised of an `id` number (unique only within a source server, not across the ecosystem), a primary `name` that the source server calls this subject, a `kind` to describe the type of subject (note that `type` is a keyword in TypeScript, thus why it was avoided here), and an optional `SubjectMetadata` object.

### `SubjectKind`

Subjects can be one of a few different kinds:

```ts
/**
 * A kind of Subject.
 */
export enum SubjectKind {
  /**
   * A participant or individual.
   */ 
  Participant = 'Participant',

  /**
   * A xenograft which was originally derived from a patient tumor.
   */
  PatientDerivedXenograft = ' Patient Derived Xenograft',

  /**
   * A cell line.
   */
  CellLine = 'Cell Line',

  /**
   * An organoid.
   */
  Organoid = 'Organoid',
}
```

### `SubjectMetadata`

Perhaps the most interesting entity of the `Subject` interface is the optional `SubjectMetadata` key:

```ts
/**
 * Metadata associated with a subject.
 */
export interface SubjectMetadata {
   /**
     * The sex of the subject as defined in caDSR CDE 6343385 v1.00.
     */
    sex: MetadataField<Sex> | null;

    /**
     * The race of the subject as defined in caDSR CDE 2192199 v1.00.
     */
    race: MetadataField<Race>[] | null;

    /**
     * The ethnicity of the subject as defined in caDSR CDE 2192217 v1.00.
     */
    ethnicity: MetadataField<Ethnicity> | null;

    /**
     * Identifiers associated with the subject as defined in caDSR CDE 6380049 v1.00.
     */
    identifiers: OwnedMetadataField<string>[] | null;

    /**
     * Unharmonized metadata fields provided by the source site.
     */
    unharmonized?: UnharmonizedMetadataFields | null;
}
```

In this proposal, all of the keys are nullable to denote when the source server does not collect or otherwise have access to the data. All of the top-level fields in the object are _required_. The `unharmonized` key is _not_ required, and we will explore that key later in this proposal.

### `MetadataField`

First, we should define the structure of `MetadataField`:

```ts
/**
 * A metadata field.
 */
export interface MetadataField<T> {
  /**
   * The value of the metadata field.
   */
  value: T;

  /**
   * The ancestors from which this field was derived. Ancestors should be
   * provided as period (`.`) delimited paths from the `metadata` key in the
   * subject response object.
   */
  ancestors?: string[];

  /**
   * A free-text comment field.
   */
  comment?: string;
}
```

This is intended to be a general purpose interface for a metadata field that transcends purely subject metadata. In other words, it is intended that sample metadata and perhaps other types of metadata will conform to this interface as well.

Notably, each metadata field has a `value`, a set of zero or more `ancestors` (ancestor metadata fields) from which this value is derived, and an optional free-text `comment`.

With respect to the convention of the `ancestors` field, as noted in the comment, the value is represented as a period (`.`) delimited path to the metadata field key from which this field was derived when considering the `metadata` object as the root.

### `OwnedMetadataField`

The `OwnedMetadataField` interface extends the `MetadataField` interface only by allowing a server to assert whether a metadata value (in this case, an identifier) is owned by the source server.

```ts
/**
 * A metadata field extended to have a concept of asserted ownership.
 */
export interface OwnedMetadataField<T> extends MetadataField<T> {
  /**
   * If known, whether or not the identifier is owned by the source server.
   */
  owned?: boolean;
}

```

This interface is used for identifiers in the proposed interface.

- If the source server considers itself to own a particular identifier, it should assert `true` here.
- If the source server knows that it is _not_ the owner of a particular identifier, it should assert `false` here.
  - If the source of the identifier is known, it is highly recommended (though not required) to include a `comment` explaining where the identifier comes from. 
- If the answer is otherwise unknown or inconclusive, the source server should exclude the `owner` key.

Obviously this introduces the possibilities for collisions (e.g., the rare case where two servers both assert they own an ID), so this is not considered authoritative and is for data collection purposes only.

### `UnharmonizedMetadataFields`

Last is the topic of the unharmonized metadata fields, which accepts zero or more arbitrarily named keys so long as the values conform to the `MetadataValue` interface:

```ts
/**
 * Unharmonized metadata fields.
 */
export interface UnharmonizedMetadataFields {
    /**
     * Any number of unspecified keys with a `MetadataField` value.
     */
    [key: string]: MetadataField<any>;
}
```

### Endpoint Details (`/subject/<id>`)

Gets the subject object (defined in the last proposed API endpoint) for the specified ID. Here, "id" refers to the `id` key in the subject object, which is a unique integer primary key that is specific to the source server's database.

Valid HTTP response codes are:

- `200`. Object was found successfully and returned in the response.
- `404`. Object was not found within the database (or, equivalently, line-level data cannot be returned for this entity).
  - Source servers which cannot show line level data should always return 404 for these requests, even if the client requests a valid `id`. The rationale is that the server should not acknowledge that _any_ ID exists to the client (for example, by returning a 200/204/403, etc).

For example, the request `/GET /subject/1` might return the following result:

```json
{
  "id": 1,
  "name": "SubjectIdentifier001",
  "kind": "Participant",
  "metadata": {
    "sex": {
      "value": "F",
      "ancestors": [
        "unharmonized.attr_sex"
      ]
    },
    "race": [
      {
        "value": "American Indian or Alaska Native",
        "ancestors": [
          "unharmonized.attr_race"
        ]
      }
    ],
    "ethnicity": {
      "value": "Not Hispanic or Latino",
      "ancestors": [
        "unharmonized.attr_ethnicity"
      ]
    },
    "identifiers": [
      {
        "value": "SubjectIdentifier001",
        "owned": true
      },
      {
        "value": "Another Identifier",
        "comment": "This identifier comes from the XYZ project.",
        "owned": false
      }
    ],
    "unharmonized": {
      "attr_sex": {
        "value": "Female"
      },
      "attr_race": {
        "value": "American Indian or Alaska Native"
      },
      "attr_ethnicity": {
        "value": "Not Hispanic or Latino"
      },
      "subject_name": {
        "value": "SubjectIdentifier001"
      },
      "field with no harmonized value": {
        "value": "Hello, world!"
      }
    }
  }
}
```

## Listing All Subjects

Next, we'll cover querying the all subjects that the server is aware of. We'll define TypeScript definitions for each request object as we did in the previous section.

The response object for the `/subject` endpoint will be as follows:

```ts
/**
 * Subjects.
 */
export interface Subjects {
  /**
   * Reports the number of subjects returned by the endpoint.
   */
  counts: TotalCount;

  /**
   * The subjects known by the server (if available).
   */
  subjects?: Subject[];
}
```

The `counts` key summarizes the number of subjects returned by the API call and is defined as follows:

```ts
/**
 * Total count of some entity as reported alongside an API call.
 */
export interface TotalCount {
    /**
     * The total number of entities returned in the API call.
     */
    total: integer;
}
```

### Endpoint Details (`/subject`)

As described above, this endpoint represents the main endpoint for retrieving line-level data from the API. 

Valid HTTP response codes are:

- `200`. Data was successfully returned within the API response.
  - Notably different from previous versions of this proposal, all endpoints return a 200 ("OK") response whether they are able to share line-level data or not. The presence of the `subjects` key with line-level information will be the differentiator between line-level returning and non line-level returning servers. 

Other notes:

- Filters would be implemented using query parameters appended to the end of the URL (`?sex=F`).

The following represents a potential API response for the `/subject` endpoint.

```json
{
  "total": 214,
  "subject": [
    {
      "id": 1,
      "name": "SubjectIdentifier001",
      "kind": "Participant",
      "metadata": {
        "sex": {
          "value": "F",
          "ancestors": [
            "unharmonized.attr_sex"
          ]
        },
        "race": [
          {
            "value": "American Indian or Alaska Native",
            "ancestors": [
              "unharmonized.attr_race"
            ]
          }
        ],
        "ethnicity": {
          "value": "Not Hispanic or Latino",
          "ancestors": [
            "unharmonized.attr_ethnicity"
          ]
        },
        "identifiers": [
          {
            "value": "SubjectIdentifier001",
            "ancestors": [
              "unharmonized.subject_name"
            ],
            "owned": true
          },
          {
            "value": "Another Identifier",
            "owned": false,
            "comment": "This identifier comes from the XYZ project."
          }
        ],
        "unharmonized": {
          "attr_ethnicity": {
            "value": "Not Hispanic or Latino"
          },
          "attr_race": {
            "value": "American Indian or Alaska Native"
          },
          "attr_sex": {
            "value": "Female"
          },
          "subject_name": {
            "value": "SubjectIdentifier001"
          },
          "field_with_no_harmonized_analog": {
            "value": "value"
          }
        }
      }
    },
    ...
  ]
}
```

### Grouping By

Often, it is desired to group samples by a particular metadata element and perform some aggregating action on them (such as count the number of subjects that have a particular value for a field). At present, the only planned aggregating action is `count`, but we have designed the namespace in such a way that other actions can be extended in the future. This proposal will only cover the count aggregating action.

First, we'll define an object that extends our `TotalCount` interface to include counts for each of the values for a particular field.

```ts
/**
 * Represents counts grouped by a particular metadata field.
 */
export interface SubjectsByCountCounts extends TotalCount {
  /**
   * A `values` object which contains keys representing the values for the
   * specified field and the number of corresponding samples with that value.
   */
  values: {
    [key: string]: integer;
  };
}
```

As the response for grouping by field values and then counting is encapsulated in the above object, we just need a simple wrapper around the full response itself.

```ts
/**
 * Subjects grouped together and counted by a particular field.
 */
export interface SubjectsByCount {
    counts: SubjectsByCountCounts;
}
```

### Endpoint Details (`/subject/by/<field>/count`)

Valid HTTP response codes are:

- `200`. The request was well-formed, the metadata element was identified, the request was processed, and the results are returned in the API response.
- `422`. The metadata element for which the client requested to group by does not exist.

Other notes:

- The field identifier is period (`.`) delimited path to a field relative to the `metadata` object. As such, if one wishes to access the unharmonized value of `attr_sex`, then the path would be `/subject/by/unharmonized.attr_sex/count`.
- The source server is only required to return metadata element values for which it has data corresponding to. For instance, if the source server does not contain any `F` values for sex, it is not required to include `"F": 0` in its response. A full list of acceptable values should be referred to the corresponding CDE.

For example, the request `GET /subject/by/sex/count` might return the following result:

```json
{
  "counts": {
    "total": 158,
    "values": {
      "F": 82,
      "M": 76
    }
  }
}
```

## Discovering Subject Metadata Fields

In the case of servers that return line-level data, supported metadata fields can be observed directly on the returned subject objects. However, this does not provide a reliable method for determining which metadata fields are supported by the server—particularly because some fields may exist on certain subjects and not others. Further, there is the outstanding issue of how supported metadata fields are discovered for non line-level returning servers. To address this gap, we'll introduce some specific discoverability endpoints.

As discussed in our meeting on September 22, 2023, we will specify two types of metadata field descriptions:

- Those which are harmonized across the federation, which will simply point to a URL describing the details of that harmonization, and
- Those which are not harmonized across the federation, which need support for things like descriptions, names, etc.

We can define both the harmonized and unharmonized metadata field descriptions. First, we'll define the harmonized description:

```ts
/**
 * A harmonized metadata field description.
 */
export interface HarmonizedMetadataFieldDescription {
  /**
   * Specify that this metadata field is an harmonized metadata field.
   */
  harmonized: true;

  /**
   * A comma (`.`) delimited path to the field's location on the `metadata`
   * objects returned by the various subject endpoints.
   */
  path: String;

  /**
   * The proper name of the standard to which this field is harmonized (defined
   * by the documentation for the CCDI metadata fields).
   */
  standard: String;

  /**
   * A URL to the CCDI documentation where the definition of this harmonized
   * field resides.
   */
  url: String;
}
```

As you can see, the `standard` type (verbatim from the formal name of the associated field in the wiki) and the `url` string (pointing to the wiki) are required for this type of metadata field description. This is in contrast to the more open-ended, unharmonized metadata field description:

```ts
/**
 * An unharmonized metadata field description.
 */
export interface UnharmonizedMetadataFieldDescription {
  /**
   * Specify that this metadata field is an unharmonized metadata field.
   */
  harmonized: false;

  /**
   * A display name for this metadata field as _suggested_ by the server (this
   * is not considered authoritative and can be ignored by the client if it so
   * chooses). This is mainly to avoid naming collisions of common fields across
   * servers.
   */
  name?: String;

  /**
   * A plain-text description of what the field represents.
   */
  description?: String;

  /**
   * A comma (`.`) delimited path to the field's location on the `metadata`
   * objects returned by the various subject endpoints.
   */
  path: String;

  /**
   * If the field is considered harmonized across the federation ecosystem, the
   * standard to which the field is harmonized.
   *
   * If the field is _not_ harmonized across the federation ecosystem, then this
   * key should be omitted.
   */
  standard?: String;

  /**
   * A url that describes more about the metadata field, if available.
   */
  url?: String;
}
```

This interface includes facilities for providing a name and an optional description while also making the URL and the standard fields optional. The above two interfaces are intended to support all metadata field descriptions, even non-subject metadata fields, as the API specification grows.

For summarizing the metadata fields, we'll define the following response object.

```ts
/**
 * A set of metadata fields associated with a subject.
 */
export interface MetadataFieldsForSubject {
  /**
   * Metadata fields supported by this API server.
   */
  fields: Array<
    UnharmonizedMetadataFieldDescription | HarmonizedMetadataFieldDescription
  >;
}
```

### Endpoint Details (`/metadata/fields/subject`)

This endpoint may only return a 200 ("OK") response listing the metadata fields supported for subjects. Below is an example API response from the fictitious server we've been building:

```json
{
  "fields": [
    {
      "harmonized": true,
      "path": "sex",
      "standard": "caDSR CDE 6343385 v1.00",
      "url": "https://github.com/CBIIT/ccdi-federation-api/wiki/Subject-Metadata-Fields#sex"
    },
    {
      "harmonized": true,
      "path": "race",
      "standard": "caDSR CDE 2192199 v1.00",
      "url": "https://github.com/CBIIT/ccdi-federation-api/wiki/Subject-Metadata-Fields#race"
    },
    {
      "harmonized": true,
      "path": "race",
      "standard": "caDSR CDE 2192217 v2.00",
      "url": "https://github.com/CBIIT/ccdi-federation-api/wiki/Subject-Metadata-Fields#ethnicity"
    },
    {
      "harmonized": true,
      "path": "identifiers",
      "standard": "caDSR CDE 6380049 v1.00",
      "url": "https://github.com/CBIIT/ccdi-federation-api/wiki/Subject-Metadata-Fields#identifiers"
    },
    {
      "harmonized": false,
      "name": "Age at Diagnosis",
      "description": "Age at first diagnosis. This field is normalized as a decimal value. If empty, the physician or research team did not indicate a value for this field.",
      "path": "unharmonized.attr_age_at_diagnosis",
      "url": "https://university.stjude.cloud/docs/genomics-platform/about-our-data/metadata-and-clinical/#clinical-and-phenotypic-information"
    },
    {
      "harmonized": false,
      "name": "Diagnosis",
      "description": "Unharmonized primary diagnosis as reported by the lab or PI upon submission of data to St. Jude Cloud.",
      "path": "unharmonized.attr_diagnosis",
      "url": "https://university.stjude.cloud/docs/genomics-platform/about-our-data/metadata-and-clinical/#clinical-and-phenotypic-information"
    },
    {
      "harmonized": false,
      "name": "Self-reported Ethnicity",
      "description": "Self-reported ethnicity. Values are normalized according to the US Census Bureau classifications.",
      "path": "unharmonized.attr_ethnicity",
      "url": "https://university.stjude.cloud/docs/genomics-platform/about-our-data/metadata-and-clinical/#clinical-and-phenotypic-information"
    },
    {
      "harmonized": false,
      "name": "Self-reported Race",
      "description": "Self-reported race. Values are normalized according to the US Census Bureau classifications.",
      "path": "unharmonized.attr_race",
      "url": "https://university.stjude.cloud/docs/genomics-platform/about-our-data/metadata-and-clinical/#clinical-and-phenotypic-information"
    },
    {
      "harmonized": false,
      "name": "Self-reported Sex",
      "description": "Self-reported sex.",
      "path": "unharmonized.attr_sex",
      "url": "https://university.stjude.cloud/docs/genomics-platform/about-our-data/metadata-and-clinical/#clinical-and-phenotypic-information"
    },
    {
      "harmonized": false,
      "name": "Diagnosis",
      "description": "Self-reported sex as reported by the investigator.",
      "path": "unharmonized.attr_sex",
      "url": "https://university.stjude.cloud/docs/genomics-platform/about-our-data/metadata-and-clinical/#clinical-and-phenotypic-information"
    }
  ]
}
```

## Initial Harmonization Standards Supported

Moved to [the wiki](https://github.com/CBIIT/ccdi-federation-api/wiki/Subject-Metadata-Fields), which is an ever-evolving document describing these fields.

**Comment by asafieva** [comment]
Here is a consideration of the aggregation operation, "by":
/subjects/by/<element>
Does this generic count has a real value when there is no other search filter added to it, as a specific study, diagnosis, etc?
GET /subjects/by/sex?diagnosis=sinus&<...>

**Reply by claymcleod** [reply]
I would argue it does have value, yes. In particular, search engines across the federation APIs are going to want to get counts by different field values (without any filters applied) when aggregating information for their own filters (e.g., data browsers where the filters are on the left side of the screen).

**Reply by melissacline** [reply]
I'm a bit confused.  Would the total aggregated count comes from the top level query
GET /subject
and then retrieve from the total object
- or -
of via the by qualifier with no serch filter
GET /subject/by
and then retrieve the total object

Will we support both?

**Reply by claymcleod** [reply]
Under the revised proposal I just submitted, it would come via the former.

**Comment by asafieva** [comment]
About using "metadata" entities embedded into "subject" entities. What value does it have compared to using attributes assigned to metadata as direct "subject" entity attributes?

Estimate of end points response payloads, as "/subjects". Do we expect that the response size will always be low? Do we need to start thinking about pagination, throttling?

**Reply by claymcleod** [reply]
> About using "metadata" entities embedded into "subject" entities. What value does it have compared to using attributes assigned to metadata as direct "subject" entity attributes?

In my personal view, adding hierarchy where possible helps bring form to the information and make it more understandable. Technically, it is possible to design an API where the entire response is a one-level JSON object, but quickly you lose the forest for the trees in terms of what information you are looking at. It is a subjective question, but one that I think this general principle should be followed: add form to the data where appropriate.

> Estimate of end points response payloads, as "/subjects". Do we expect that the response size will always be low? Do we need to start thinking about pagination, throttling?

Pagination, should likely be considered, though I think it could be pushed to a further iteration of the API (pre-public, of course, but keeping our focus on the content of the API right now seems prudent to me). For the throttling, I view that as an implementation detail left to the implementers, not something that should be dictated by the API spec.

**Reply by melissacline** [reply]
+++1!  I have a data resource that returns a flat JSON object, and while there are technical reasons for doing that, it sure complicates my software development!

**Comment by jingchunzhu** [comment]
regarding the size of payload might exceed limit, how about using POST.

**Reply by asafieva** [reply]
Sorry for the confusion, I meant the size of _response_ restrictions. What if someone requested to return all the entities of some type from an endpoint? Is it OK?

**Reply by claymcleod** [reply]
My feeling is that it is okay for pre-production. We do need to worry about those details before releasing to the public, but I have been thinking that we would ignore them for now while we determine the content of the response.

**Comment by claymcleod** [comment]
- I have updated `subjects` to `subject` as we discussed in the meeting.
- It's still an open question as to whether this approach balances the concerns of `subject` vs `participant` correctly—let me know what you all think.
- For now, I have kept my proposed `harmonized` field underneath each attribute because this is the most salient way I can visualize it. If others have opinions on how this could be structured, I would really like to see what the actual response object might look like so we could consider the pros and cons of each.

**Comment by claymcleod** [comment]
I have made the following updates to the proposal:

- I have begun adding TypeScript definitions to concretely define all of the responses. This has been a really helpful exercise, particularly as the proposed responses change rapidly and need updating and re-rendering for the discussion. I hope others find this helpful as well.
- Harmonized fields are now prioritized as top-level keys in the `metadata` field of responses. There is an `unharmonized` key that allows arbitrary key-value pairs to be added for site-specific metadata.
- The `identifiers` metadata field was changed from harmonizing to caDSR CDE 6867052 v1.00 (“Specimen Source Subject Identifier”) to caDSR CDE 6380049 v1.00 (“Subject Identifier for the Study”), which is not retired.
- The concept of ownership of identifiers was also added to the response objects. In particular, servers may now assert whether they feel they own a particular ID. Obviously this introduces the possibilities for collisions (e.g., the rare case where two servers both assert they own an ID), so this is not considered authoritative and is for data collection purposes only.
- Metadata fields can now optionally specify one or more `ancestors` from which the field was derived. This is to support tracing of what non-harmonized fields the harmonized fields are derived from. As was pointed out in our discussion, sometimes this can be multiple fields, so we have made this an array.
  - Harmonized fields are allowed to omit the `ancestors` key, indicating it is possible that the harmonized value did not have an analog unharmonized value.
  - Conversely, unharmonized values can stand alone with no harmonized counterpart. 
- We've introduced a new concept of a "metadata field path", which is a period (".") delimited path to the location of a metadata field when considering the `metadata` object as the root. This is useful in all sorts of contexts, not the least of which is the concept of ancestry of fields.
- The 204 ("No Content") response has been removed from the `/subject` endpoint in favor of always returning a 200 (`OK`) response. Servers that cannot return line-level data should simply omit the `subjects` key. This solves the problem of how to get the total number of subjects from a server as well.

Outstanding questions:

- Currently, the `metadata` field is omittable (as indicated by the `?` after the key name in the TypeScript definition). What this means is that you are allowed to return a subject without any metadata attached. Is this a desired behavior?
- At present, I've made the decision that all values of metadata must be cast to a `string`. This simplifies parsing, as the client should always know what type to expect, but sacrifices explicitness. The more I think about it, the more I'd like to allow for other native JSON types here, as clients will have to deal with parsing and casting values anyways (for instance, if the value is an integer).
- In the proposal, I introduce the concept of metadata field descriptions. Of note, there is a key called `standard`: it's currently an open question as to whether this should represent a plain-text description of the standard or if this should be an identifier that can further be explored within the API (say, at an endpoint like `/metadata/fields/<name>` that lists the permissible values and also values collected by the server).
- Last, I've introduced the concept of `unharmonized` metadata underneath the associated key in the `metadata` object. It strikes me though that this may not be the right word, as the values are likely harmonized (albiet only within one server). I wonder if there is a better word to indicate "unharmonized across the entire ecosystem"?

**Comment by grugna** [comment]
@claymcleod It felt to short to ask it at the end of the call so I am just writing here. Talking about `SubjectIdentifierMetadataField` you mentioned something about a link concept? Could you tell me a little more about that.

I am asking / am proposing to have some sort of metadata about who owns the identifiers because for instance in our case we don't own the USI that are the IDs we are using. Specially in a future context where we will have CPI/MPI I think we should have a source_id of some sort that we can match to the data loaded in the master index. In our case it would be COG to be specific.

**Reply by claymcleod** [reply]
Yeah it's a good point. I do have a `comment?` field there, which any free-text can be stored, but it might be worth adding some more structured facilities such as a `owner_url` where you can find out more from the owner. Would any link be allowed there? Would only links to other CCDI federation APIs be allowed there? It may also be a better long-term solution to have the MPI assign identifiers for ID minters, and our API just simply tracks those and responds back with them—of course, that might be a solution for down the road.

**Reply by grugna** [reply]
Definitely something we don't need now and we can tackle later.

**Comment by claymcleod** [comment]
Based on our discussion today, I have updated the proposal in the following ways:

- `Individual` was changed to `Participant` for a `SubjectKind`.
- Metadata fields now have two types: (a) one for harmonized fields, which simply point to a description in the wiki and (b) one for unharmonized fields, which provides facilities for a few more pieces of information.
- As I caught during the meeting, it was an error that `subjects` was not an optional key in the `/subject` response, so I have now made that optional.

**Comment by grugna** [comment]
@claymcleod One more thing I want to bring up here. Just like the /sample discussion we need to have a way to say 403 or cohort too small. At least for the time being we would be able to return only the `total` and not the `subject` portion of the response. However if the total is lower than 5 by our policy the existing endpoint currently return -1. This is to minimize the ability from the user to reconstruct the line level data by creating many cohorts of 1 subject.

This may be a question for the larger group and not just the API endpoint, but do we want to enforce a minimum cohort size in the response or other people are ok with displaying a cohort of 1?
